Shallow clone using git to build images.

calavera opened the pull request:
This will make the cloning way faster by downloading only the most recent commit in the repo's history, rather than downloading the whole history.

Signed-off-by: David Calavera david.calavera@gmail.com

calavera committed:
Shallow clone using git to build images.

Signed-off-by: David Calavera <david.calavera@gmail.com>

icecrime commented:
Oh, that's interesting!

There are a few tests failing:

```
--- FAIL: TestBuildApiLowerDockerfile (0.06s)
	docker_api_containers_test.go:434: Build failed: received status != 200 OK: 500 Internal Server Error
		"Error trying to use git: exit status 128 (Cloning into '/var/lib/docker/tmp/docker-build-git250078911'...\nfatal: dumb http transport does not support --depth\n)\n"
--- FAIL: TestBuildApiBuildGitWithF (0.05s)
	docker_api_containers_test.go:465: Build failed: received status != 200 OK: 500 Internal Server Error
		"Error trying to use git: exit status 128 (Cloning into '/var/lib/docker/tmp/docker-build-git346876690'...\nfatal: dumb http transport does not support --depth\n)\n"
--- FAIL: TestBuildApiDoubleDockerfile (0.04s)
	docker_api_containers_test.go:496: Build failed: received status != 200 OK: 500 Internal Server Error
```

Other than that:
* I think it requires design review as it *might* be breaking some (weird) use cases
* I'm +1 as long as it's properly documented, because I'd be surprised people rely today on getting the whole history.

calavera commented:
As far as I know, there is no way to hijack the build process to manipulate the repo, is there? there are high chances that I'm mistaken :smile:

icecrime commented:
@calavera Well, it's the context, so technically you can!

```
$ docker build -t foo https://github.com/icecrime/PoC_-12502.git
[...]
Step 4 : RUN git log -2
 ---> Running in 0ba84c005406
commit 6097f214ed67c5bd874304829cc647c60fe7fd5e
Author: Arnaud Porterie <arnaud.porterie@docker.com>
Date:   Fri Apr 17 16:05:28 2015 -0700

    Dockerfile

    Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>

commit 24613673426527b23d4275b7ba5f6d7d43e26fdd
Author: Arnaud Porterie <arnaud.porterie@docker.com>
Date:   Fri Apr 17 16:04:24 2015 -0700

    bar

    Signed-off-by: Arnaud Porterie <arnaud.porterie@docker.com>
```

calavera commented:
oh, yeah, you're right. I'm going to close this PR, it might cause more trouble than good.

ewindisch commented:
@calavera users desiring what @icecrime demonstrated could simply do a 'git fetch' to pull the rest of their history. I think that's more reasonable than having everyone do a deep pull.

calavera commented:
@ewindisch I think I agree with you, but I'm not sure how to keep this change backwards compatible, or if  we event want to. I can restore the branch and open it again for further discussion if people really want this behavior.

ewindisch commented:
I don't think the current behavior is so common a use-case that the workaround won't be sufficient. Of course, we can have that conversation with anyone that might deem to protest the change!

calavera commented:
Okay, reopened for further discussion. Now I "just" need to figure out how to hook git's smart http protocol in those tests that fail :grimacing:

calavera committed:
Do not try to shallow git history when the protocol doesn't allow it.

This only happens with the old git http dumb protocol, but that's what we use in our integration tests.
We check the Content-Type header advertised in http requests to make sure the http transport is the git smart transport:

See this commit as a reference:
https://github.com/git/git/commit/4656bf47fca857df51b5d6f4b7b052192b3b2317

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
@icecrime, @ewindisch I ended up adding a check to make sure old git http transports still work, although there should not be infrastructure out there using it. It's easier to keep our integration tests using dumb git http than trying to migrate to the smart protocol.

calavera committed:
Test that we set the right arguments for git cloning depending on the transport.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
I've added some tests to make sure we still send the right arguments. Let me know what you think.

LK4D4 commented:
LGTM

tiborvass commented:
I'm +1 for this

tianon commented:
+1 here too -- if someone needs more explicit control over how the repo is
cloned, they should clone it directly instead of letting Docker clone it
for them IMO :+1:

icecrime commented:
LGTM

icecrime commented:
I'm pushing this to `3-needs-docs-review` because I'm pretty sure there is some documentation impact.

calavera committed:
Document the extra `depth` argument in git contexts.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Remove duplicated git clone logic.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
@icecrime, @LK4D4 do you mind to review the code again? #12670 made me realize that the logic is duplicated between daemon and client. I've extracted it to a utils function.

jfrazelle commented:
LGTM

moxiegirl commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -636,12 +636,13 @@ refer to any of the files in the context. For example, your build can use
-in a temporary directory on your local host. Then, this clone is sent to
-the Docker daemon as the context. Local clones give you the ability to
-access private repositories using local user credentials, VPN's, and so forth.
+The `URL` parameter can specify the location of a Git repository,
```
The `URL` parameter can specify the location of a Git repository; the repository acts as the build context. The system recursively clones the repository and its submodules using a `git clone --depth 1 --recursive` command. This command runs in a temporary directory on your local host.
After the command succeeds, the directory is sent to the Docker daemon as the context.
Local clones give you the ability to access private repositories using
local user credentials, VPN's, and so forth.

calavera commented:
@moxiegirl I've fixed the documentation, thanks for your corrections!

moxiegirl commented:
LGTM
